File: admin/app/Actions/Extractors/DetermineIdType.php
<?php

namespace App\Actions\Extractors;

use App\Enums\IdType;

final class DetermineIdType
{
    /**
     * @param  OcrItem[]  $items
     * @param  array<string, IdType>  $labelDictionary
     */
    public function execute(array $items, array $labelDictionary): IdType
    {
        /** @var array<string, int> $votes */
        $votes = [];
        $labels = array_keys($labelDictionary);

        foreach ($items as $item) {
            $matchedLabel = $this->matchLabel($item->text, $labels);

            if ($matchedLabel !== null) {
                $type = $labelDictionary[$matchedLabel]->value;
                $votes[$type] = ($votes[$type] ?? 0) + 1;
            }
        }

        if (isset($votes[IdType::SimModern->value]) || isset($votes[IdType::SimSixLine->value]) || isset($votes[IdType::SimOld->value])) {
            unset($votes[IdType::Sim->value]);
        }

        if (empty($votes)) {
            return IdType::Ktp;
        }

        arsort($votes);

        return IdType::from(array_key_first($votes));
    }

    /**
     * @param  string[]  $labels
     */
    private function matchLabel(string $text, array $labels): ?string
    {
        $text = trim($text);

        if ($text === '') {
            return null;
        }

        // Exact match (fast path)
        if (in_array($text, $labels, true)) {
            return $text;
        }

        // Fuzzy match via Levenshtein
        $fuzzyMatch = $this->fuzzyMatch($text, $labels);

        if ($fuzzyMatch !== null) {
            return $fuzzyMatch;
        }

        // Prefix detection for merged label+value strings (e.g. "Gol.DarahB", "Kewarganegaraan:WNI")
        return $this->prefixMatch($text, $labels);
    }

    /**
     * @param  string[]  $labels
     */
    private function fuzzyMatch(string $text, array $labels): ?string
    {
        $bestLabel = null;
        $bestDistance = PHP_INT_MAX;
        $textLower = mb_strtolower($text);

        foreach ($labels as $label) {
            $distance = levenshtein($textLower, mb_strtolower($label));

            if ($distance < $bestDistance) {
                $bestDistance = $distance;
                $bestLabel = $label;
            }
        }

        if ($bestLabel === null) {
            return null;
        }

        $maxLen = max(strlen($text), strlen($bestLabel));

        if ($maxLen <= 3) {
            return $bestDistance <= 1 ? $bestLabel : null;
        }

        return $maxLen > 0 && ($bestDistance / $maxLen) < 0.3 ? $bestLabel : null;
    }

    /**
     * @param  string[]  $labels
     */
    private function prefixMatch(string $text, array $labels): ?string
    {
        $bestLabel = null;
        $bestLen = 0;

        foreach ($labels as $label) {
            if (strlen($label) <= 3 || strlen($label) >= strlen($text)) {
                continue;
            }

            $prefix = substr($text, 0, strlen($label));
            $distance = levenshtein(mb_strtolower($prefix), mb_strtolower($label));
            $maxLen = max(strlen($prefix), strlen($label));

            if ($maxLen > 0 && ($distance / $maxLen) < 0.3 && strlen($label) > $bestLen) {
                $bestLabel = $label;
                $bestLen = strlen($label);
            }
        }

        return $bestLabel;
    }
}

File: ocr/extractors/determine_id_type.py
from __future__ import annotations

from rapidfuzz.distance import Levenshtein

from .fuzzy_match import passes_threshold
from .id_type import IdType
from .ocr_item import OcrItem


class DetermineIdType:
    SIM_HEADER_LABELS = frozenset(
        {"SURAT IZIN MENGEMUDI", "DRIVING LICENSE", "INDONESIA"}
    )

    def execute(
        self,
        items: list[OcrItem],
        label_dictionary: dict[str, IdType],
    ) -> IdType:
        votes: dict[str, int] = {}
        saw_sim_header = False
        labels = list(label_dictionary.keys())

        for item in items:
            matched_label = self._match_label(item.text, labels)

            if matched_label is None:
                continue

            if matched_label in self.SIM_HEADER_LABELS:
                saw_sim_header = True
                continue

            type_value = label_dictionary[matched_label].value
            votes[type_value] = votes.get(type_value, 0) + 1

        if not votes:
            return IdType.SIM_MODERN if saw_sim_header else IdType.KTP

        max_count = max(votes.values())

        for key, value in votes.items():
            if value == max_count:
                return IdType(key)

        return IdType.KTP

    def _match_label(self, text: str, labels: list[str]) -> str | None:
        text = text.strip()

        if not text:
            return None

        if text in labels:
            return text

        fuzzy = self._fuzzy_match(text, labels)
        if fuzzy is not None:
            return fuzzy

        return self._prefix_match(text, labels)

    def _fuzzy_match(self, text: str, labels: list[str]) -> str | None:
        best_label = None
        best_distance = float("inf")
        text_lower = text.lower()

        for label in labels:
            distance = Levenshtein.distance(text_lower, label.lower())

            if distance < best_distance:
                best_distance = distance
                best_label = label

        if best_label is None:
            return None

        if passes_threshold(best_distance, len(text), len(best_label)):
            return best_label

        return None

    def _prefix_match(self, text: str, labels: list[str]) -> str | None:
        best_label = None
        best_len = 0

        for label in labels:
            label_len = len(label)
            if label_len <= 3 or label_len >= len(text):
                continue

            prefix = text[:label_len]
            distance = Levenshtein.distance(prefix.lower(), label.lower())

            if passes_threshold(distance, len(prefix), label_len) and label_len > best_len:
                best_label = label
                best_len = label_len

        return best_label
